ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/LivenessFilter.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server;
+
+import static java.util.Objects.requireNonNull;
+
+import io.confluent.ksql.execution.streams.IRoutingFilter;
+import io.confluent.ksql.rest.server.HeartbeatAgent.HostStatus;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.kafka.streams.state.HostInfo;
+
+public class LivenessFilter implements IRoutingFilter {
+
+  private final Optional<HeartbeatAgent> heartbeatAgent;
+
+  public LivenessFilter(final Optional<HeartbeatAgent> heartbeatAgent) {
+    this.heartbeatAgent = requireNonNull(heartbeatAgent, "heartbeatAgent");
+  }
+
+  @Override
+  public boolean filter(final HostInfo hostInfo, final String storeName, final int partition) {
+    if (heartbeatAgent.isPresent()) {
+      final Map<HostInfo, HostStatus> hostStatus = heartbeatAgent.get().getHostsStatus();
```

Comment:
You're making a copy of the map in `getHostsStatus` at every filtering that's happening.  This is a small nit, but you might consider having the underlying map be an ImmutableMap so you can just return it since this will be theoretically happening at a much higher rate than updates.

Reply:
If the map is not a `ConcurrentMap` anymore, but rather an `ImmutableMap`, I will have to synchronize every access on it or make it volatile and copy on every write, correct? If this is indeed better, this seems like a larger change better suitable for a different PR? What do you think?

Reply:
Tracked in issue: https://github.com/confluentinc/ksql/issues/4393

Reply:
All you have to do is use an AtomicReference and set it to an ImmutableMap, which is threadsafe.  Every time you get updates on all of the statuses, you build a new ImmutableMap and set the atomic reference to it.

You don't have to make the change here since this is currently threadsafe.  Just an optimization we can do.